_posts/2017-07-03-Devoxx-pl.md

```diff
@@ -165,4 +165,203 @@ With tools like [Split](https://www.split.io) and [LaunchDarkly](https://launchd
+
+{% highlight Java %}
+jshell> CompletableFuture.runAsync(()->{while(true){}})
+$7 ==> java.util.concurrent.CompletableFuture@26aa12dd[Not completed]
+
+jshell> $7.copy()
+$8 ==> java.util.concurrent.CompletableFuture@711f39f9[Not completed]
+
+jshell> $8.cancel(true)
+$9 ==> true
+
+jshell> $8
+$8 ==> java.util.concurrent.CompletableFuture@711f39f9[Completed exceptionally]
+
+jshell> $7
+$7 ==> java.util.concurrent.CompletableFuture@26aa12dd[Not completed, 1 dependents]
+{% endhighlight %}
+
+A new `ProcessHandle` will be able to get information and control processes.
+
+### Bits and pieces
+
+The underscore will become a keyword, so assigning a value to _ does not work.
+This is probably a feature for the future where _ will be used for matching arguments of any type.
+Assigning a value to __ will keep on working.
+In Java 8 default methods were added to interfaces, in 9 they can be private.
+Property files will support UTF 9 and there is already Java 9 support in several IDEs.
+There will be several changes to improve `String` performance, by for example using a more space-efficient internal representation for Strings.
+Javadoc will get an improved search, HTML5 compliance and more info on the module where the class or interface comes from.
+The use of agents will be more flexible, a process can attach an agent to itself and a jar can contain multiple agents.
+
+### Modules
+The Java Platform Modules System (JPMS) allows modularization for Java applications.
+A module can define dependent modules with _requires_, to provide an API the _exports_ key word is used.
+To give access to everyone the _opens_ keyword can be used.
+
+There is a method `getAccessable`, beware using it for determining if a module is usable.
+Actually it just returns the value of `setAccessable`, a toggle that you have to set yourself.
+
+To make a smooth transition to the JPMS, any jar on the classpath will become an automatic module.
+By default '--illegal-access=permit' is the default mode for JDK 9, allowing modules access to all automatic modules.
+
+As a migration strategy Oleg proposes to wait for dependencies to modularize before modularizing yourself.
+Otherwise you might need to modularize twice to align with dependencies.
+
+Java 9 with maven is complicated and not yet fully integrated, Gradle releases fixes more often and currently supports more features.
+
+A multi release jar containing multiple versions for the same file, in the same jar, is a new feature that should be used with caution.
+
+### Garbage collection
+The G1 Garbage Collector (G1GC) will become the default in Java 9.
+Previous Garbage Collector's were not so scalable nor predictable.
+The G1GC promises a more scalable and more predictable system with few modification options.
+
+By default 1/4 of the physical ram will be allocated to the heap, unless the size is specified with the `-mx` flag.
+Due to its new heap division system it might run into problems with large chunks of data.
```

Comment:
Due to its new heap division system **comma** it might run into problems with large chunks of data.